### ISSUE TITLE
Support Clang8+

### ISSUE THREAD

**Issue by vorj** [fixstars/clpy#213]
LLVM/Clang 8.0.0 was released. We should support it.

Tasks are below:
- [x] Investigate about affected APIs
- [x] Add workaround codes when we need

**Comment by vorj** [comment]
```sh-session
$ clang++ --version
clang version 8.0.0 (tags/RELEASE_800/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /path/to/.local/bin
$ pwd
/path/to/clpy/headercvt
$ make use_cxx11_abi:=1
clang++ -std=c++11 -D_GLIBCXX_USE_CXX11_ABI=1 -Wall -Wextra -O3 -pedantic-errors headercvt.cpp -lclangTooling -lclangFrontendTool -lclangFrontend -lclangDriver -lclangSerialization -lclangCodeGen -lclangParse -lclangSema -lclangStaticAnalyzerFrontend -lclangStaticAnalyzerCheckers -lclangStaticAnalyzerCore -lclangAnalysis -lclangEdit -lclangAST -lclangLex -lclangBasic -lclang `llvm-config --libs --system-libs` -fno-rtti -o headercvt
/tmp/headercvt-174118.o: In function `main':
headercvt.cpp:(.text+0x384): undefined reference to `clang::vfs::getRealFileSystem()'
headercvt.cpp:(.text+0x3ab): undefined reference to `clang::tooling::ClangTool::ClangTool(clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::shared_ptr<clang::PCHContainerOperations>, llvm::IntrusiveRefCntPtr<clang::vfs::FileSystem>)'
clang-8: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:4: recipe for target 'headercvt' failed
make: *** [headercvt] Error 1
$ cd /path/to/.local/lib
$ for x in *.a; do nm -C $x | grep clang::vfs::getRealFileSystem; done
$
```

🤔

**Comment by vorj** [comment]
Ughhhh, I found it why `headercvt` can't build with Clang 8 !

```shell-session
$ echo $PATH
/path/to/.local/bin
$ echo $CPLUS_INCLUDE_PATH

```

Clang 8 read headers for system-installed old clang...

**Comment by vorj** [comment]
- Headercvt needs no fix for clang update.
- Ultima needs some fixes for below breaking changes:
    - `clang::PredefinedExpr` discards some members: `getIdentTypeName` , `getIdentType`
    - Identifier `StringRef` not found

**Comment by vorj** [comment]
#228 fixed this issue, so I'm closing it.

**Comment by vorj** [comment]
I confirmed that the ultima built with Clang8 doesn't work correctly. I'm reopening it...

**Comment by vorj** [comment]
I haven't completed investigation so I don't know why ultima built using clang 8 doesn't work correctly, but the ultima stop output without completing her work.
We can't say "ClPy supports Clang8" in this situation, so I will revert 314a6bc for now.

**Comment by vorj** [comment]
LLVM 8.0.1 was released. However, ultima built with it doesn't work correctly, either.

**Comment by vorj** [comment]
LLVM 9.0.0 has been released. I fixed ultima codes for some breaking changes, but the ultima built with clang 9.0.0 doesn't work correctly, either.

**Comment by vorj** [comment]
https://github.com/fixstars/clpy/blob/cc72c544a17c823ab9af443366f2191b00746ba1/clpy/core/carray.pxi#L178

It should be `if proc.returncode != 0 or len(errstream) > 0:` .
When we replace the `and` to `or` , we can get `SIGSEGV` from ultima built with newer clang.
I will fix this.

**Comment by vorj** [comment]
Note: Clang 8+ has new AST node type, `clang::ConstantExpr` . In the code of current ultima, we have gotten the value of integer (non-type) template parameter as follows:

```
clang::dyn_cast<clang::IntegerLiteral>(template_type->begin()->getAsExpr())->getValue().getLimitedValue();
```

However, in Clang 8+, `template_type->begin()->getAsExpr()` is not `clang::IntegerLiteral` but `clang::ConstantExpr` which has `clang::IntegerLiteral` as sub expression.
So, current ultima built with Clang 8+ have met her demise.

This problem will be fixed on 7f8cd72 .
